packages/apollo-server-codegen/src/IR/Types.ts

```diff
@@ -0,0 +1,123 @@
+import { FieldNode, TypeNode, Kind } from "graphql";
+import { Translatable, Translator } from "../Translators";
+import { TypelessObjectDefinition, SELECTION_OFFSET } from "./Objects";
+
+export interface TypeDefinition extends Translatable {
+  translate(translator: Translator, nullable?: boolean): string;
+}
+
+export class NamedType implements TypeDefinition {
+  constructor(public name: string) {}
+
+  public translate(translator: Translator, nullable: boolean = true) {
+    return translator.translateNamedType(this, nullable);
+  }
+}
+
+export class NonNullType implements TypeDefinition {
+  constructor(public base: TypeDefinition) {}
+
+  public translate(translator: Translator) {
+    return translator.translateNonNullType(this);
+  }
+}
+
+export class ListType implements TypeDefinition {
+  constructor(public base: TypeDefinition) {}
+
+  public translate(translator: Translator, nullable: boolean = true) {
+    return translator.translateListType(this, nullable);
+  }
+}
+
+export const findRootType = (t: TypeDefinition): string => {
```

Comment:
Hmm, I thought this was already a utility in either our code base or in graphql-js. I can't seem to find it off hand but maybe @JakeDawkins or @martijnwalraven remembers where it is?

One thing to note, we try to not use `instanceof` because it can break across duplicate module boundaries even if the package is of the same version. We are going to be changing this in graphql-js at some point, but for the time being, can we use either duck typing or predicate functions for these utilities to minimize the spreading of `instanceof` in the codebase?

Reply:
Got rid of the instance of checks. This is specifically finding the root type for my IR types, so I don't think you'd have it anywhere else.